File: iOS 8 by Tutorials/Objective-C/Chapter 22 - Intermediate Live Rendering/Intermediate Live Rendering/WatchControl/WatchControl/WatchView.m
/*
 * Copyright (c) 2014 Razeware LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#import "WatchView.h"

@interface WatchView()
@property (nonatomic, strong) NSTimer *timer;
@end

@implementation WatchView

- (instancetype)initWithFrame:(CGRect)frame {
	self = [super initWithFrame:frame];
	if (self) {
		[self commonInit];
	}
	return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
	self = [super initWithCoder:coder];
	if (self) {
		[self commonInit];
	}
	return self;
}

- (void) commonInit
{
	//Initialize here.
}

- (void)layoutSubviews
{
	[super layoutSubviews];
	[self createClockFace];
}

- (void)updateLayerProperties {
	if (self.ringProgress == 0) {
		self.ringLayer.strokeEnd = self.ringProgress / 60.0;
	}
	
	if (self.backgroundImageLayer) {
		if (self.backgroundImage)
		{
			self.backgroundImageLayer.contents = (id)self.backgroundImage.CGImage;
		}
	}
	
	if (self.backgroundLayer) {
		self.backgroundLayer.fillColor = self.backgroundLayerColor.CGColor;
	}
	
	if (self.enableAnalogDesign) {
		[self setHideAnalogDesign:NO];
		[self setHideDigitalDesign:YES];
		
		if (self.enableClockSecondHand) {
			[self setHideSecondClockHand:NO];
			[self setHideRingLayer:YES];
		} else if (!self.enableClockSecondHand) {
			[self setHideSecondClockHand:YES];
			[self setHideRingLayer:NO];
		} else {
			[self setHideRingLayer:YES];
			[self setHideSecondClockHand:YES];
		}
	} else {
		[self setHideDigitalDesign:NO];
		[self setHideAnalogDesign:YES];
	}
	
	if (self.enableColorBackground) {
		[self setHideImageBackground:YES];
	} else {
		[self setHideImageBackground:NO];
	}
}

#pragma mark Methods to create Clock Face
- (void)createClockFace
{
	[self layoutBackgroundLayer];
	[self layoutBackgroundImageLayer];
	
	if (self.enableAnalogDesign) {
		[self createAnalogClock];
	} else {
		[self createDigitalClock];
	}
	
	[self updateLayerProperties];
}

- (void)createAnalogClock
{
	[self layoutMinuteHandLayer];
	[self layoutHourHandLayer];
	
	if (self.enableClockSecondHand) {
		[self layoutSecondHandLayer];
	} else {
		[self layoutWatchRingLayer];
	}
}

- (void)createDigitalClock
{
	[self layoutHourMinuteSecondText];
	[self layoutAmPmText];
	[self layoutWeekDayText];
}

#pragma mark - Creating Base Layers
- (void)layoutBackgroundLayer
{
	if (!self.backgroundLayer)
	{
		self.backgroundLayer = [CAShapeLayer layer];
		[self.layer addSublayer:self.backgroundLayer];
		CGRect insetBounds = CGRectInset(self.bounds, self.lineWidth / 2.0, self.lineWidth / 2.0);
		UIBezierPath *innerPath = [UIBezierPath bezierPathWithOvalInRect:insetBounds];
		self.backgroundLayer.path = innerPath.CGPath;
		self.backgroundLayer.fillColor = self.backgroundLayerColor.CGColor;
		self.backgroundLayer.lineWidth = self.lineWidth;
		self.backgroundLayer.strokeColor = [UIColor colorWithWhite:0.5 alpha:0.5].CGColor;
	}
	self.backgroundLayer.frame = self.bounds;
}

- (void)layoutWatchRingLayer
{
	if (!self.ringLayer) {
		self.ringLayer = [CAShapeLayer layer];
		[self.layer addSublayer:self.ringLayer];
		CGRect rect = CGRectInset(self.bounds, self.ringThickness / 2.0, self.ringThickness / 2.0);
		UIBezierPath *path = [UIBezierPath bezierPathWithOvalInRect:rect];
		self.ringLayer.transform = CATransform3DMakeRotation(-(M_PI / 2), 0, 0, 1);
		self.ringLayer.strokeColor = self.ringColor.CGColor;
		self.ringLayer.path = path.CGPath;
		self.ringLayer.fillColor = nil;
		self.ringLayer.lineWidth = self.ringThickness;
	}
	self.ringLayer.frame = self.layer.bounds;
}

- (void)layoutBackgroundImageLayer
{
	if (!self.backgroundImageLayer)
	{
		CAShapeLayer *maskLayer = [CAShapeLayer layer];
		CGRect insetBounds = CGRectInset(self.bounds, self.lineWidth + 3.0, self.lineWidth + 3.0);
		UIBezierPath *innerPath = [UIBezierPath bezierPathWithOvalInRect:insetBounds];
		maskLayer.path = innerPath.CGPath;
		maskLayer.fillColor = [[UIColor blackColor] CGColor];
		maskLayer.frame = self.bounds;
		[self.layer addSublayer:maskLayer];
		
		self.backgroundImageLayer = [CAShapeLayer layer];
		[self.layer addSublayer:self.backgroundImageLayer];
		self.backgroundImageLayer.frame = self.bounds;
		self.backgroundImageLayer.contentsGravity = kCAGravityResizeAspectFill;
		self.backgroundImageLayer.mask = maskLayer;
	}
}

#pragma mark - Creating the different clock hands
- (void)layoutSecondHandLayer
{
	if (!self.secondHandLayer) {
		self.secondHandLayer = [self createClockHandWithAnchorPoint:CGPointMake(1.0, 1.0) withHandLength:20.0 withHandWidth:4.0 withHandAlpha:1.0 withHandColor:self.secondHandColor];
		[self.layer addSublayer:self.secondHandLayer];
	}
}

- (void)layoutMinuteHandLayer
{
	if (!self.minuteHandLayer) {
		self.minuteHandLayer = [self createClockHandWithAnchorPoint:CGPointMake(1.0, 1.0) withHandLength:26.0 withHandWidth:12.0 withHandAlpha:1.0 withHandColor:self.minuteHandColor];
		[self.layer addSublayer:self.minuteHandLayer];
	}
}

- (void)layoutHourHandLayer
{
	if (!self.hourHandLayer) {
		self.hourHandLayer = [self createClockHandWithAnchorPoint:CGPointMake(1.0, 1.0) withHandLength:65.0 withHandWidth:12.0 withHandAlpha:1.0 withHandColor:self.hourHandColor];
		[self.layer addSublayer:self.hourHandLayer];
	}
}

- (void)layoutHourMinuteSecondText
{
	if (!self.hourMinuteSecondLayer) {
		self.hourMinuteSecondLayer = [self createTextLayerWithString:@"00:00:00" withFont:self.bounds.size.height/8.5 withOffset:self.bounds.size.height/2];
		[self.layer addSublayer:self.hourMinuteSecondLayer];
	}
}

- (void)layoutAmPmText
{
	if (!self.ampmLayer) {
		self.ampmLayer = [self createTextLayerWithString:@"am" withFont:self.bounds.size.height/13.0 withOffset:self.bounds.size.height/1.6];
		[self.layer addSublayer:self.ampmLayer];
	}
}

- (void)layoutWeekDayText
{
	if (!self.weekdayLayer) {
		self.weekdayLayer = [self createTextLayerWithString:@"Thursday" withFont:self.bounds.size.height/13.0 withOffset:self.bounds.size.height];
		[self.layer addSublayer:self.weekdayLayer];
	}
}

#pragma mark - Methods for hidding various CALayer components
- (void)setHideImageBackground:(BOOL)willHide
{
	if (self.backgroundImageLayer) {
		self.backgroundImageLayer.hidden = willHide;
	}
}

- (void)setHideSecondClockHand:(BOOL)willHide
{
	if (self.secondHandLayer) {
		self.secondHandLayer.hidden = willHide;
	}
}

- (void)setHideRingLayer:(BOOL)willHide
{
	if (self.ringLayer) {
		self.ringLayer.hidden = willHide;
	}
}

- (void)setHideAnalogDesign:(BOOL)willHide
{
	if (self.secondHandLayer) {
		self.secondHandLayer.hidden = willHide;
	}
	if (self.hourHandLayer) {
		self.hourHandLayer.hidden = willHide;
	}
	if (self.minuteHandLayer) {
		self.minuteHandLayer.hidden = willHide;
	}
	if (self.ringLayer) {
		self.ringLayer.hidden = willHide;
	}
}

- (void)setHideDigitalDesign:(BOOL)willHide
{
	if (self.hourMinuteSecondLayer) {
		self.hourMinuteSecondLayer.hidden = willHide;
	}
	if (self.ampmLayer) {
		self.ampmLayer.hidden = willHide;
	}
	if (self.weekdayLayer) {
		self.weekdayLayer.hidden = willHide;
	}
}

#pragma mark - Helper Methods to create CAShapeLayers, and CATextLayers
//Create clock hands of various sizes, alpha, color, anchor point etc.
- (CAShapeLayer *)createClockHandWithAnchorPoint:(CGPoint)anchorPoint withHandLength:(CGFloat)handLength withHandWidth: (CGFloat)handWidth withHandAlpha:(CGFloat)handAlpha withHandColor:(UIColor *)handColor
{
	CAShapeLayer *handLayer = [CAShapeLayer layer];
	UIBezierPath *path = [UIBezierPath bezierPath];
	[path moveToPoint:CGPointMake(1.0, handLength)];
	[path addLineToPoint:CGPointMake(1.0, (self.bounds.size.height / 2.0))];
	
	handLayer.bounds = CGRectMake(0.0, 0.0, 1.0, self.bounds.size.height / 2.0);
	handLayer.anchorPoint = anchorPoint;
	handLayer.position = CGPointMake(CGRectGetMidX(self.bounds), CGRectGetMidY(self.bounds));
	handLayer.lineWidth = handWidth;
	handLayer.opacity = handAlpha;
	handLayer.strokeColor = handColor.CGColor;
	handLayer.path = path.CGPath;
	handLayer.lineCap = kCALineCapRound;
	return handLayer;
}

//Create and returns a CATextLayer with a specific font and offset
- (CATextLayer *)createTextLayerWithString:(NSString *)string withFont: (CGFloat)fontSize withOffset: (CGFloat)offset
{
	
	CATextLayer *layer = [[CATextLayer alloc] init];
	[layer setFont:@"HelveticaNeue-Light"];
	[layer setFontSize:fontSize];
	[layer setFrame:CGRectMake(0.0, 0.0, self.bounds.size.width / 2.0, self.bounds.size.height / 2.0)];
	layer.position = CGPointMake(CGRectGetMidX(self.bounds), offset);
	[layer setString:string];
	[layer setAlignmentMode:kCAAlignmentCenter];
	[layer setForegroundColor:[[UIColor whiteColor] CGColor]];
	[layer setContentsScale:[[UIScreen mainScreen] scale]];
	
	return layer;
}

#pragma mark - Methods to start the clock ticking
- (void)startTimeWithTimeZone:(NSString *)string
{
	[self endTime];
	//Set the current time zone user has selected.
	self.currentTimeZone = string;
	
	if (self.enableAnalogDesign)
	{
		self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(animateAnalogClock) userInfo:nil repeats:YES];
	}
	else
	{
		self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(animateDigitalClock) userInfo:nil repeats:YES];
	}
}

- (void)endTime
{
	[self.timer invalidate];
	self.timer = nil;
}

- (void)animateAnalogClock
{
	//Get today's time.
	NSDate *now = [NSDate date];
	
	//Create a date formatter, and set the time zone selected by the user.
	NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
	[dateFormatter setTimeZone:[NSTimeZone timeZoneWithName:self.currentTimeZone]];
	[dateFormatter setDateFormat:@"hh:mm:ss"];
	
	//Extract the hour, minute, and second from the date string.
	NSArray *dateComponents = [self grabDateComponentsWithString:[dateFormatter stringFromDate:now]];
	
	//Convert string to integers
	NSInteger hours = [dateComponents[0] integerValue];
	NSInteger minutes = [dateComponents[1] integerValue];
	NSInteger seconds = [dateComponents[2] integerValue];
	
	NSInteger minutesIntoDay = hours * 60 + minutes;
	float percentageMinutesIntoDay = minutesIntoDay / (12.0 * 60.0);
	float percentageMinutesIntoHour = (float)minutes / 60.0;
	float percentageSecondsIntoMinute = (float)seconds / 60.0;
	
	//rotates the second hand, if second hand is enabled.
	if (self.enableClockSecondHand)
	{
		self.secondHandLayer.transform = CATransform3DMakeRotation((M_PI * 2) * percentageSecondsIntoMinute, 0, 0, 1);
	}
	else
	{
		self.ringLayer.strokeEnd = seconds / 60.0;
	}
	
	if (self.minuteHandLayer) {
		self.minuteHandLayer.transform = CATransform3DMakeRotation((M_PI * 2) * percentageMinutesIntoHour, 0, 0, 1);
	}
	
	if (self.hourHandLayer) {
		self.hourHandLayer.transform = CATransform3DMakeRotation((M_PI * 2) * percentageMinutesIntoDay, 0, 0, 1);
	}
}

- (void)animateDigitalClock
{
	//Get today's time.
	NSDate *now = [NSDate date];
	
	//Create a date formatter, and set the time zone selected by the user.
	NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
	[dateFormatter setTimeZone:[NSTimeZone timeZoneWithName:self.currentTimeZone]];
	[dateFormatter setDateFormat:@"hh:mm:ss:a:EEEE"];
	
	//Extract the hour, minute, and second from the date string.
	NSArray *dateComponents = [self grabDateComponentsWithString:[dateFormatter stringFromDate:now]];
	
	//Convert string to integers
	NSInteger hours = [dateComponents[0] integerValue];
	NSInteger minutes = [dateComponents[1] integerValue];
	NSInteger seconds = [dateComponents[2] integerValue];
	NSString *ampm = [dateComponents[3] lowercaseString];
	NSString *weekday = dateComponents[4];
	
	NSString *hourminutesecondString = [NSString stringWithFormat:@"%02ld:%02ld:%02ld", (long)hours, (long)minutes, (long)seconds];
	if (self.hourMinuteSecondLayer) {
		[self.hourMinuteSecondLayer setString:hourminutesecondString];
	}
	if (self.ampmLayer) {
		[self.ampmLayer setString:ampm];
	}
	
	if (self.weekdayLayer) {
		[self.weekdayLayer setString:weekday];
	}
}

#pragma mark - Copying Setting

- (void)copyClockSettingsWithWatchView:(WatchView *)clock
{
	self.enableAnalogDesign = clock.enableAnalogDesign;
	self.enableClockSecondHand = clock.enableClockSecondHand;
	self.enableColorBackground = clock.enableColorBackground;
	
	self.ringThickness = clock.ringThickness;
	self.ringColor = clock.ringColor;
	self.ringProgress = clock.ringProgress;
	self.hourHandColor = clock.hourHandColor;
	self.minuteHandColor = clock.minuteHandColor;
	self.secondHandColor = clock.secondHandColor;
	self.backgroundImage = clock.backgroundImage;
	self.lineWidth = clock.lineWidth;
	self.backgroundLayerColor = clock.backgroundLayerColor;
	self.currentTimeZone = clock.currentTimeZone;
}

#pragma mark - Utilities
//Had some problems getting the actual time from a time zone, so i'm manually stripping out the hour, minute, and seconds. NSDate was outputting the wrong thing/
//array[0] = hour, //array[1] = minute, // array[2] = seconds
- (NSArray *) grabDateComponentsWithString: (NSString *)dateString
{
	NSArray *dateArray = [dateString componentsSeparatedByString:@":"];
	return dateArray;
}

@end

File: iOS 8 by Tutorials/Swift/Chapter 22 - Intermediate Live Rendering/WatchControl-Starter/WatchControlSwift/WatchView.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

import UIKit
import QuartzCore

@IBDesignable
class WatchView: UIView {
  
  @IBInspectable var enableClockSecondHand: Bool = false {
    didSet { updateLayerProperties() }
  }

  @IBInspectable var enableColorBackground: Bool = false {
    didSet { updateLayerProperties() }
  }
  //Ring Layer - Circular style progress bar that ticks clockwise.
  var ringLayer: CAShapeLayer!
  @IBInspectable var ringThickness: CGFloat = 2.0
  @IBInspectable var ringColor: UIColor = UIColor.blueColor()
  @IBInspectable var ringProgress: CGFloat = 45.0/60 {
    didSet { updateLayerProperties() }
  }

  //Base color background layer
  var backgroundLayer: CAShapeLayer!
  @IBInspectable var backgroundLayerColor: UIColor = UIColor.lightGrayColor() {
    didSet { updateLayerProperties() }
  }
  @IBInspectable var lineWidth: CGFloat = 1.0

  //Image background layer
  var backgroundImageLayer: CALayer!
  @IBInspectable var backgroundImage: UIImage? {
    didSet  { updateLayerProperties() }
  }

  //Watch second "hand"
  var secondHandLayer: CAShapeLayer!
  @IBInspectable var secondHandColor: UIColor = UIColor.redColor()

  //Watch minute "hand"
  var minuteHandLayer: CAShapeLayer!
  @IBInspectable var minuteHandColor: UIColor = UIColor.whiteColor()

  //Watch hour "hand"
  var hourHandLayer: CAShapeLayer!
  @IBInspectable var hourHandColor: UIColor = UIColor.whiteColor()

  //What makes the clock ticky ticky :]
  var timer = NSTimer()
  var currentTimeZone: String = "Asia/Singapore"

  override init(frame: CGRect) {
    super.init(frame: frame)
    commonInit()
  }

  required init(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    commonInit()
  }

  func commonInit() {
    //Initialize whatever here.
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    createClockFace()
  }

  func createClockFace() {
    //First we create the base background layer.
    layoutBackgroundLayer()
    //Then we layout the background image if there is one.
    layoutBackgroundImageLayer()

    createAnalogClock()

    //If we haven't changed any of the inspectable properties.
    updateLayerProperties()
  }

  func createAnalogClock() {
    //Lays out the minute hand.
    layoutMinuteHandLayer()
    //Lays out the hour hand.
    layoutHourHandLayer()
    
    if enableClockSecondHand == true {
      //Lays out the second hand.
      layoutSecondHandLayer()
    } else {
      //Next we lay out the ring indicator.
      layoutWatchRingLayer()
    }
  }

  //The base circular layer.
  func layoutBackgroundLayer() {
    if backgroundLayer == nil {
      backgroundLayer = CAShapeLayer()
      layer.addSublayer(backgroundLayer)
      
      let rect = CGRectInset(bounds, lineWidth / 2.0, lineWidth / 2.0)
      let path = UIBezierPath(ovalInRect: rect)
      
      backgroundLayer.path = path.CGPath
      backgroundLayer.fillColor = backgroundLayerColor.CGColor
      backgroundLayer.lineWidth = lineWidth
      backgroundLayer.strokeColor = UIColor(white: 0.5, alpha: 0.3).CGColor
    }
    backgroundLayer.frame = layer.bounds
  }

  //Creates the watch's seconds indicator in a cirular form.
  func layoutWatchRingLayer() {
    if ringProgress == 0 {
      if ringLayer != nil {
        ringLayer.strokeEnd = 0.0
      }
    }
    
    if ringLayer == nil {
      ringLayer = CAShapeLayer()
      let dx = ringThickness / 2.0
      let rect = CGRectInset(bounds, dx, dx)
      let path = UIBezierPath(ovalInRect: rect)
      ringLayer.transform = CATransform3DMakeRotation(CGFloat(-(M_PI/2)), 0, 0, 1)
      ringLayer.strokeColor = ringColor.CGColor
      ringLayer.path = path.CGPath;
      ringLayer.fillColor = nil;
      ringLayer.lineWidth = ringThickness
      layer.addSublayer(ringLayer)
    }
    
    ringLayer.frame = layer.bounds
  }

  //Creates the background image layer
  func layoutBackgroundImageLayer() {
    if backgroundImageLayer == nil {
      let maskLayer = CAShapeLayer()
      let dx = lineWidth + 3.0
      let insetBounds = CGRectInset(layer.bounds, dx, dx)
      let innerPath = UIBezierPath(ovalInRect: insetBounds)
      maskLayer.path = innerPath.CGPath
      maskLayer.fillColor = UIColor.blackColor().CGColor
      maskLayer.frame = layer.bounds
      layer.addSublayer(maskLayer)
      
      backgroundImageLayer = CAShapeLayer()
      backgroundImageLayer.mask = maskLayer
      backgroundImageLayer.contentsGravity = kCAGravityResizeAspectFill
      layer.addSublayer(backgroundImageLayer)
      backgroundImageLayer.frame = layer.bounds
    }
  }

  //Creates the second hand layer
  func layoutSecondHandLayer() {
    if secondHandLayer == nil {
      secondHandLayer = createClockHand(CGPointMake(1.0, 1.0), handLength: 20.0, handWidth: 4.0, handAlpha: 1.0, handColor: secondHandColor)
      layer.addSublayer(secondHandLayer)
    }
  }

  //Creates the minute hand layer
  func layoutMinuteHandLayer() {
    if minuteHandLayer == nil {
      minuteHandLayer = createClockHand(CGPointMake(1.0, 1.0), handLength: 26.0, handWidth: 12.0, handAlpha: 1.0, handColor: minuteHandColor)
      layer.addSublayer(minuteHandLayer)
                      }
  }

  func layoutHourHandLayer() {
    if hourHandLayer == nil {
      hourHandLayer = createClockHand(CGPointMake(1.0, 1.0), handLength: 65.0, handWidth:12.0, handAlpha: 1.0, handColor: hourHandColor)
      layer.addSublayer(hourHandLayer)
    }
  }

  func layoutHourMinuteSecondText() {
    //TODO:Implement
  }

  func layoutAmPmText() {
    //TODO: Implement
  }

  func layoutWeekDayText() {
    //TODO:Implement
  }

  //MARK: Property Observer Function
  func updateLayerProperties() {
    //Updates the ring progress.
    if ringLayer != nil {
      ringLayer.strokeEnd = ringProgress / 60.0
    }
    
    //Updates the image layer.
    if backgroundImageLayer != nil {
      if let image = backgroundImage {
        backgroundImageLayer.contents = image.CGImage
      }
    }
    
    if backgroundLayer != nil {
      backgroundLayer.fillColor = backgroundLayerColor.CGColor
    }
    
    // If user toggles between second hand and ring progress indicator
    if enableClockSecondHand == true {
      setHideSecondClockHand(false)
      setHideRingLayer(true)
    } else if enableClockSecondHand == false {
      setHideSecondClockHand(true)
      setHideRingLayer(false)
    } else {
      setHideRingLayer(true)
      setHideSecondClockHand(true)
    }
    
    //Handles toggling between color background and image background.
    if enableColorBackground == true {
      setHideImageBackground(true)
    } else {
      setHideImageBackground(false)
    }
  }

  //MARK: Hiding Components
  func setHideImageBackground(willHide: Bool) {
    if backgroundImageLayer != nil {
      backgroundImageLayer.hidden = willHide
    }
  }

  func setHideSecondClockHand(willHide: Bool) {
    if secondHandLayer != nil {
      secondHandLayer.hidden = willHide
    }
  }

  func setHideRingLayer(willHide: Bool) {
    if ringLayer != nil {
      ringLayer.hidden = willHide
    }
  }

  func setHideAnalogDesign(willHide: Bool) {
    //TODO: Implement
  }

  func setHideDigitalDesign(willHide: Bool) {
    //TODO: Implement
  }

  //MARK: Helper Methods
  //Helper function that creates clock hands of different length and sizes or what not!
  func createClockHand(anchorPoint: CGPoint, handLength: CGFloat, handWidth: CGFloat, handAlpha: CGFloat, handColor: UIColor) -> CAShapeLayer {
    let handLayer = CAShapeLayer()
    let path = UIBezierPath()
    path.moveToPoint(CGPointMake(1.0, handLength))
    path.addLineToPoint(CGPointMake(1.0, bounds.size.height / 2.0))
    handLayer.frame = CGRectMake(0.0, 0.0, 1.0, bounds.size.height / 2.0)
    handLayer.anchorPoint = anchorPoint
    handLayer.position = CGPointMake(CGRectGetMidX(bounds), CGRectGetMidY(bounds))
    handLayer.lineWidth = handWidth
    handLayer.opacity = Float(handAlpha)
    handLayer.strokeColor = handColor.CGColor
    handLayer.path = path.CGPath
    handLayer.lineCap = kCALineCapRound
    handLayer.contentsScale = UIScreen.mainScreen().scale
    
    return handLayer
  }

  //Splits date components seperated by ":"
  func grabDateComponents(dateString: String) -> [String] {
    let dateArray = dateString.componentsSeparatedByString(":")
    return dateArray
  }

  //MARK: Start and Stop Time Controls
  func startTimeWithTimeZone(timezone: String) {
    //Set the current time zone selected
    endTime()
    currentTimeZone = timezone
    timer = NSTimer.scheduledTimerWithTimeInterval(1.0, target: self, selector: "animateAnalogClock", userInfo: nil, repeats: true)
  }

  func endTime() {
    timer.invalidate()
  }

  //The stuff that makes the clock just tick
  func animateAnalogClock() {
    //Get today's time.
    let now = NSDate()
    //Create a date formatter, and set the time zone selected by the user.
    let dateFormatter = NSDateFormatter()
    dateFormatter.timeZone = NSTimeZone(name: currentTimeZone)
    dateFormatter.dateFormat = "hh:mm:ss"
    //Extract the hour, minute, and second from the date string
    let dateComponents = grabDateComponents(dateFormatter.stringFromDate(now))
    //Convert string to integers
    let hours = dateComponents[0].toInt()
    let minutes = dateComponents[1].toInt()
    let seconds = dateComponents[2].toInt()
    
    //Calculate percentage to rotate clock hands by.
    let minutesIntoDay = CGFloat(hours!) * 60.0 + CGFloat(minutes!);
    let pminutesIntoDay = CGFloat(minutesIntoDay) / (12.0 * 60.0)
    let minutesIntoHour = CGFloat(minutes!) / 60.0
    let secondsIntoMinute = CGFloat(seconds!) / 60.0
		
    if enableClockSecondHand == true {
      if secondHandLayer != nil {
        secondHandLayer.transform = CATransform3DMakeRotation(CGFloat(M_PI * 2.0) * secondsIntoMinute, 0, 0, 1)
      }
    } else {
      if ringLayer != nil {
        ringProgress = CGFloat(seconds!)
      }
    }
    
    if minuteHandLayer != nil {
      minuteHandLayer.transform = CATransform3DMakeRotation(CGFloat(M_PI * 2) * minutesIntoHour, 0, 0, 1)
    }
    
    if hourHandLayer != nil {
      hourHandLayer.transform = CATransform3DMakeRotation(CGFloat(M_PI * 2) * pminutesIntoDay, 0, 0, 1)
    }
  }

  func animateDigitalClock() {
    //TODO: Implement
  }

  func copyClockSettings(clock: WatchView) {
    //TODO: Implement
  }
}




